File: cpp/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/PolygonProcessor.cpp
#include <cmath>

#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/common/dataStructures/BinaryTreeNode.h"
#include "vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/PolygonProcessor.h"
static bool pointOnSegment(const Vertex2D& p, const Vertex2D& a, const Vertex2D& b)
{
    double cross = (p.y-a.y)*(b.x-a.x) - (p.x-a.x)*(b.y-a.y);
    if (std::fabs(cross) > 1e-12) return false;
    double dot = (p.x-a.x)*(b.x-a.x) + (p.y-a.y)*(b.y-a.y);
    if (dot < 0) return false;
    double len2 = (b.x-a.x)*(b.x-a.x) + (b.y-a.y)*(b.y-a.y);
    return dot <= len2;
}

Polygon2D* PolygonProcessor::polygon2DSimplify(const Polygon2D& pol2DIn, double epsilon, bool copy)
{
    (void)epsilon;
    Polygon2D* out = new Polygon2D();
    for (long int i = 0; i < out->loops.size(); ++i) delete out->loops[i];
    out->loops.clear();
    for (long int i = 0; i < pol2DIn.loops.size(); ++i) {
        _Polygon2DContour* src = pol2DIn.loops.get(i);
        _Polygon2DContour* dst = new _Polygon2DContour();
        for (long int j = 0; j < src->vertices.size(); ++j) {
            Vertex2D p = src->vertices[j];
            if (copy) dst->vertices.add(Vertex2D(p.x,p.y,p.color.r(),p.color.g(),p.color.b()));
            else dst->vertices.add(p);
        }
        out->loops.add(dst);
    }
    return out;
}

signed char PolygonProcessor::isPointInsidePolygon2D(const Vertex2D& point, const java::ArrayList<Vertex2D>& polygon)
{
    bool inside = false;
    long int n = polygon.size();
    if (n < 3) return -1;
    for (long int i = 0, j = n - 1; i < n; j = i++) {
        Vertex2D vi = polygon.get(i);
        Vertex2D vj = polygon.get(j);
        if (pointOnSegment(point, vj, vi)) return 0;
        bool intersect = ((vi.y > point.y) != (vj.y > point.y)) &&
                         (point.x < (vj.x - vi.x) * (point.y - vi.y) / ((vj.y - vi.y) + 1e-30) + vi.x);
        if (intersect) inside = !inside;
    }
    return inside ? 1 : -1;
}

bool PolygonProcessor::contourInsidePolygon(const java::ArrayList<Vertex2D>& contour, const java::ArrayList<Vertex2D>& mainPolygon)
{
    for (long int i = 0; i < contour.size(); ++i) {
        signed char r = isPointInsidePolygon2D(contour.get(i), mainPolygon);
        if (r == 1) return true;
        if (r == -1) return false;
    }
    return false;
}

void PolygonProcessor::classifyContourHoles(Polygon2D* polygon)
{
    if (polygon == 0) return;
    for (long int i = 0; i < polygon->loops.size(); ++i) polygon->loops[i]->setExteriorContour(0);
    BinaryTreeNode<_Polygon2DContour*>* root = new BinaryTreeNode<_Polygon2DContour*>(0);
    polygon->setHeadNode(root);
    for (long int i = 0; i < polygon->loops.size(); ++i) {
        _Polygon2DContour* a = polygon->loops[i];
        for (long int j = 0; j < polygon->loops.size(); ++j) {
            if (i == j) continue;
            _Polygon2DContour* b = polygon->loops[j];
            if (contourInsidePolygon(a->vertices, b->vertices)) {
                a->setExteriorContour(b);
                break;
            }
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/PolygonProcessor.ts
import { BinaryTreeNode } from "../../../../common/dataStructures/BinaryTreeNode.js";
import { ProcessingElement } from "../../../../processing/ProcessingElement.js";
import { Vertex2D } from "../../element/Vertex2D.js";
import { Polygon2D } from "../../surface/polygon/Polygon2D.js";
import { _Polygon2DContour } from "../../surface/polygon/_Polygon2DContour.js";

/** Non-recursive Ramer--Douglas--Peucker and polygon-contour utilities. */
export class PolygonProcessor extends ProcessingElement {
    private static readonly EPSILON = 1e-18;

    public static polygon2DSimplify(pol2DIn: Polygon2D, epsilon: number, copy: boolean): Polygon2D {
        const simplified = new Polygon2D();
        simplified.loops.length = 0;
        // Java's public method always passes true to its private helper. Keep
        // that historical behavior, even though its API documents `copy`.
        void copy;
        for (const contour of pol2DIn.loops)
            simplified.loops.push(this.polygon2DContourSimplify(contour, epsilon, true));
        return simplified;
    }

    public static classifyContourHoles(polygon: Polygon2D): void {
        for (const contour of polygon.loops) contour.fleetingFlag = false;
        const head = new BinaryTreeNode<_Polygon2DContour>(null as unknown as _Polygon2DContour);
        polygon.setHeadNode(head);
        for (let i = 0; i < polygon.loops.length; i++) {
            const contour = polygon.loops[i]!;
            const contained = [contour];
            for (let j = i + 1; j < polygon.loops.length; j++) {
                const candidate = polygon.loops[j]!;
                if (!candidate.fleetingFlag && this.contourInsidePolygon(contour.vertices, candidate.vertices))
                    contained.push(candidate);
            }
            contained.sort((a, b) => b.compareTo(a));
            this.insertListInBinaryTree(contained, head);
        }
    }

    /** Returns true when the contour lies inside the supplied polygon contour. */
    public static contourInsidePolygon(contour: readonly Vertex2D[], mainPolygon: readonly Vertex2D[]): boolean {
        for (const point of contour) {
            const result = this.isPointInsidePolygon2D(point, mainPolygon);
            if (result === 1) return true;
            if (result === -1) return false;
        }
        return false;
    }

    /** Returns 1 inside, -1 outside, and 0 on the boundary. */
    public static isPointInsidePolygon2D(point: Vertex2D, polygon: readonly Vertex2D[]): number {
        if (polygon.length < 3) return -1;
        let isInside = false;
        for (let i = 0; i < polygon.length; i++) {
            const current = polygon[i]!;
            const next = polygon[(i + 1) % polygon.length]!;
            let temp = current.y - next.y;
            if (Math.abs(temp) < this.EPSILON && Math.abs(point.y - (next.y + temp / 2)) < this.EPSILON) {
                if ((next.x - point.x) * (point.x - current.x) >= 0) return 0;
            }
            if (point.y < current.y !== point.y < next.y) {
                temp = ((next.x - current.x) * (point.y - current.y)) / (next.y - current.y) + current.x;
                if (Math.abs(point.x - temp) < this.EPSILON) return 0;
                if (point.x < temp) isInside = !isInside;
            }
        }
        return isInside ? 1 : -1;
    }

    private static polygon2DContourSimplify(
        contour: _Polygon2DContour,
        epsilon: number,
        copy: boolean,
    ): _Polygon2DContour {
        const simplified = new _Polygon2DContour();
        const count = contour.vertices.length;
        if (count < 3) {
            for (const point of contour.vertices) simplified.vertices.push(copy ? this.copyVertex(point) : point);
            return simplified;
        }
        const flags = new Array<number>(count).fill(0);
        const missingNodes = [count - 1, 0];
        while (missingNodes.length > 0) {
            const ind0 = missingNodes.pop()!;
            const ind1 = missingNodes.pop()!;
            flags[ind0] = 1;
            flags[ind1] = 1;
            if (ind1 - ind0 > 1) {
                const [indFar, distance] = this.getFarthestNodeToLine(contour, ind0, ind1);
                if (distance > epsilon) missingNodes.push(indFar, ind0, ind1, indFar);
            }
        }
        for (let i = 0; i < count; i++) {
            if (flags[i] === 1) {
                const point = contour.vertices[i]!;
                simplified.vertices.push(copy ? this.copyVertex(point) : point);
            }
        }
        return simplified;
    }

    private static getFarthestNodeToLine(contour: _Polygon2DContour, ind0: number, ind1: number): [number, number] {
        const first = contour.vertices[ind0]!;
        const last = contour.vertices[ind1]!;
        const deltaY = last.y - first.y;
        let nx: number;
        let ny: number;
        if (deltaY < 0.00001 && deltaY > -0.00001) {
            nx = 0;
            ny = 1;
        } else {
            const slope = -(last.x - first.x) / deltaY;
            const normalLength = Math.sqrt(1 + slope * slope);
            nx = 1 / normalLength;
            ny = slope / normalLength;
        }
        let maximumDistance = 0;
        let farthest = ind0 + 1;
        for (let i = ind0 + 1; i < ind1; i++) {
            const point = contour.vertices[i]!;
            const distance = Math.abs((point.x - first.x) * nx + (point.y - first.y) * ny);
            if (distance > maximumDistance) {
                maximumDistance = distance;
                farthest = i;
            }
        }
        return [farthest, maximumDistance];
    }

    private static insertListInBinaryTree(list: _Polygon2DContour[], head: BinaryTreeNode<_Polygon2DContour>): void {
        if (list.length === 0) return;
        const [containingNode, containingLevel] = this.findContainingContourNode(list[0]!, head);
        this.insertListInBinaryTreeNode(list, containingNode, containingLevel);
    }

    private static insertListInBinaryTreeNode(
        list: _Polygon2DContour[],
        containingNode: BinaryTreeNode<_Polygon2DContour>,
        containingLevel: number,
    ): void {
        let level = containingLevel + 1;
        if (level % 2 === 0) list[0]!.setExteriorContour(containingNode.getData());
        let lastChild = containingNode.getChild();
        if (lastChild === null) {
            lastChild = new BinaryTreeNode(list[0]!);
            containingNode.setChild(lastChild);
        } else {
            while (lastChild.getSibling() !== null) lastChild = lastChild.getSibling()!;
            const child = new BinaryTreeNode(list[0]!);
            lastChild.setSibling(child);
            lastChild = child;
        }
        list[0]!.fleetingFlag = true;
        for (let i = 1; i < list.length; i++) {
            level++;
            const contour = list[i]!;
            if (level % 2 === 0) contour.setExteriorContour(lastChild.getData());
            const child = new BinaryTreeNode(contour);
            lastChild.setChild(child);
            contour.fleetingFlag = true;
            lastChild = child;
        }
    }

    private static findContainingContourNode(
        contour: _Polygon2DContour,
        head: BinaryTreeNode<_Polygon2DContour>,
    ): [BinaryTreeNode<_Polygon2DContour>, number] {
        let current = head;
        let level = 0;
        while (true) {
            let child = current.getChild();
            let found = false;
            while (child !== null) {
                if (this.contourInsidePolygon(contour.vertices, child.getData().vertices)) {
                    current = child;
                    level++;
                    found = true;
                    break;
                }
                child = child.getSibling();
            }
            if (!found) return [current, level];
        }
    }

    private static copyVertex(point: Vertex2D): Vertex2D {
        return new Vertex2D(point.x, point.y, point.color.r(), point.color.g(), point.color.b());
    }
}
